doc/scripts/gen_mon_command_api.py

```diff
@@ -0,0 +1,224 @@
+        if simple:
+            if self.type == "CephChoices" and self.strings:
+                return val
+            elif self.type == "CephString" and self.name != 'who':
+                return 'my_' + self.name
+            else:
+                return CmdParam.bash_example[self.type]
+        else:
+            return '--{}={}'.format(self.name, val)
+
+    def __str__(self):
+        return '* **{}** (*{}*)'.format(self.name(), self.type())
+
+
+class CmdCommand(object):
+    def __init__(self, sig, desc, module=None, perm=None, flags=0, poll=None):
+        self.sig = [s for s in sig if isinstance(s, str)]
+        self.params = sorted([CmdParam(**s) for s in sig if not isinstance(s, str)],
+                             key=lambda p: p.req, reverse=True)
+        self.help = desc
+        self.module = module
+        self.perm = perm
+        self.flags = Flags(flags)
+        self.needs_overload = False
+
+    def prefix(self):
+        return ' '.join(self.sig)
+
+    def mk_params(self):
+        return '\n'.join([f.help() for f in self.params])
+
+    def mk_doc_string(self):
+        elems = []
+        if self.help:
+            elems.append(textwrap.fill(self.help))
+
+        elems.append(self.mk_bash_example())
+
+        if self.params:
+            elems.append('Parameters:')
+            elems.append(self.mk_params())
+
+        if self.module:
+            elems.append('Ceph Module:')
+            elems.append('* *{}*'.format(self.module))
+
+        if self.perm:
+            elems.append('Required Permissions:')
+            elems.append('* *{}*'.format(self.perm))
+
+        if str(self.flags):
+            elems.append('Command Flags:')
+            elems.append('* *{}*'.format(str(self.flags)))
+
+        return '\n\n'.join(elems)
+
+    def is_reasonably_simple(self):
+        if len(self.params) > 3:
+            return False
+        if any(p.n for p in self.params):
+            return False
+        return True
+
+    def mk_bash_example(self):
+        simple = self.is_reasonably_simple()
+        line = ' '.join(['ceph', self.prefix()] + [p.mk_bash_example(simple) for p in self.params])
+        return "Example command:\n\n.. code-block:: bash\n\n    " + line
+
+    def __str__(self):
+        return command_template.format(
+            prefix=self.prefix(),
+            prefix_section_chars=''.join(['^'] * len(self.prefix())),
+            doc_string=self.mk_doc_string()
+        )
+
+
+def mk_sigs(all):
+    sigs = [CmdCommand(**e) for e in all]
+    sigs = [s for s in sigs if 'hidden' not in s.flags]
+    sigs = sorted(sigs, key=lambda f: f.sig)
+
+    print(header)
+    print('\n'.join([str(s) for s in sigs]))
+    print(footer)
+
+
+if __name__ == '__main__':
+    script_dir = os.path.dirname(os.path.realpath(__file__))
+    commands = json.loads(check_output([sys.executable, script_dir + '/../../src/script/gen_static_command_descriptions.py']))
+    mk_sigs(commands)
```

Comment:
I'm curious, instead of directly generating the text output from the python module, have you considered using a template engine (like Jinja) so for this module you can generate multiple outputs for the same data:
```
gen_mon_command_api.py --template rst.j2  # RST
gen_mon_command_api.py --template html.j2  # HTML
gen_mon_command_api.py --template py.j2  # Python code
```

Reply:
most of the code here is really just about the sphix docs. I feel we gain little by introducing a template engine. Maybe that#s something to investigate in a future enhancement.

Reply:
Looks more understandable now with the template stuff. Thanks!

Reply:
only if you know jinja2 :smile: